### PR TITLE
Incremental context bound verification

### PR DESCRIPTION
**Pull request by lucasccordeiro**
This PR adds support for incremental context-bound verification for multi-threaded C programs.

I have run this PR using different values to `context-bound-step` and `max-context-bound` using the SV-COMP benchmarks. The best results were obtained using the following parameters:

```
--incremental-cb
--context-bound-step 5 (starting at 2, 7, 12)
--max-context-bound 15
--no-por
--unwind 8
```

As a result, I have obtained the following results:

Before this PR:

```
Statistics:           1082 Files
  correct:             967
    correct true:      182
    correct false:     785
  incorrect:            21
    incorrect true:     14
    incorrect false:     7
  unknown:              94
  Score:               589 (max: 1344)
```

After this PR:

```
Statistics:           1082 Files
  correct:             876
    correct true:      102
    correct false:     774
  incorrect:             5
    incorrect true:      2
    incorrect false:     3
  unknown:             201
  Score:               866 (max: 1344)
```

### CODE REVIEW

File: src/esbmc/esbmc_parseoptions.cpp [hunk 1]
```diff
@@ -1114,6 +1117,69 @@ int esbmc_parseoptionst::doit_incremental()
...
+    char str[5];
+    snprintf(str, sizeof(str), "%d", context_bound);
+    opts.set_option("context-bound", str);
+    std::cout << "\n*** Context bound number ";
+    std::cout << context_bound;
+    std::cout << " ***\n";
+
+    if(do_base_case(opts, goto_functions, unwind))
+      return true;
+    else if(!do_forward_condition(opts, goto_functions, unwind))
```
Comment (mikhailramalho) [inline src/esbmc/esbmc_parseoptions.cpp]: no need for an `else` here.

File: src/esbmc/esbmc_parseoptions.cpp [hunk 2]
```diff
@@ -1114,6 +1117,69 @@ int esbmc_parseoptionst::doit_incremental()
...
+
+  // Get the context bound increment
+  unsigned int context_bound_inc =
+    strtoul(cmdline.getval("context-bound-step"), nullptr, 10);
+
+  // Get the unwind bound
+  unsigned unwind = atoi(opts.get_option("unwind").c_str());
+
+  bool unknown = true;
+
```
Comment (mikhailramalho) [inline src/esbmc/esbmc_parseoptions.cpp]: No need for the variable `unknown` in this code.

File: src/esbmc/esbmc_parseoptions.cpp [hunk 3]
```diff
@@ -1666,7 +1732,8 @@ void esbmc_parseoptionst::help()
        "location\n"
        " --ssa-no-sliced              do not print the sliced SSAs\n"
        " --ssa-full-names             print SSAs with full variable names\n"
-       " --smt-formula-only           only show SMT formula (not supported by "
+       " --smt-formula-only           only show SMT formula (not supported "
+       "by "
```
Comment (mikhailramalho) [inline src/esbmc/esbmc_parseoptions.cpp]: No need to change this line.

File: src/esbmc/esbmc_parseoptions.cpp [hunk 4]
```diff
@@ -1681,18 +1748,21 @@ void esbmc_parseoptionst::help()
        " --symex-ssa-trace            print generated SSA during symbolic "
        "execution\n"
        " --ssa-trace                  print SSA during SMT encoding\n"
-       " --ssa-smt-trace              print generated SMT during SMT encoding\n"
+       " --ssa-smt-trace              print generated SMT during SMT "
+       "encoding\n"
```
Comment (mikhailramalho) [inline src/esbmc/esbmc_parseoptions.cpp]: No need to change this line.

File: src/esbmc/esbmc_parseoptions.cpp [hunk 5]
```diff
@@ -1681,18 +1748,21 @@ void esbmc_parseoptionst::help()
...
        " --show-goto-value-sets       show value-set analysis for the goto "
        "functions\n"
-       " --show-symex-value-sets      show value-set analysis during symbolic "
+       " --show-symex-value-sets      show value-set analysis during "
+       "symbolic "
```
Comment (mikhailramalho) [inline src/esbmc/esbmc_parseoptions.cpp]: No need to change this line.

File: src/esbmc/esbmc_parseoptions.cpp [hunk 6]
```diff
@@ -1681,18 +1748,21 @@ void esbmc_parseoptionst::help()
...
        "execution\n"
 
        "\nFront-end options\n"
        " -I path                      set include path\n"
        " -D macro                     define preprocessor macro\n"
        " --preprocess                 stop after preprocessing\n"
        " --no-inlining                disable inlining function calls\n"
-       " --full-inlining              perform full inlining of function calls\n"
+       " --full-inlining              perform full inlining of function "
+       "calls\n"
```
Comment (mikhailramalho) [inline src/esbmc/esbmc_parseoptions.cpp]: No need to change this line.

File: src/esbmc/esbmc_parseoptions.cpp [hunk 7]
```diff
@@ -1704,11 +1774,13 @@ void esbmc_parseoptionst::help()
        "code\n"
        " --little-endian              allow little-endian word-byte "
        "conversions\n"
-       " --big-endian                 allow big-endian word-byte conversions\n"
+       " --big-endian                 allow big-endian word-byte "
+       "conversions\n"
```
Comment (mikhailramalho) [inline src/esbmc/esbmc_parseoptions.cpp]: No need to change this line.

File: src/esbmc/esbmc_parseoptions.cpp [hunk 8]
```diff
@@ -1704,11 +1774,13 @@ void esbmc_parseoptionst::help()
...
        " --16, --32, --64             set width of machine word (default is "
        "64)\n"
        " --version                    show current ESBMC version and exit\n"
-       " --witness-output filename    generate the verification result witness "
+       " --witness-output filename    generate the verification result "
+       "witness "
```
Comment (mikhailramalho) [inline src/esbmc/esbmc_parseoptions.cpp]: No need to change this line.

File: src/esbmc/esbmc_parseoptions.cpp [hunk 9]
```diff
@@ -1717,7 +1789,8 @@ void esbmc_parseoptionst::help()
        " --i386-macos                 set MACOS/I386 architecture\n"
        " --ppc-macos                  set PPC/I386 architecture\n"
        " --i386-linux                 set Linux/I386 architecture\n"
-       " --i386-win32                 set Windows/I386 architecture (default)\n"
+       " --i386-win32                 set Windows/I386 architecture "
+       "(default)\n"
```
Comment (mikhailramalho) [inline src/esbmc/esbmc_parseoptions.cpp]: No need to change this line.

File: src/esbmc/esbmc_parseoptions.cpp [hunk 10]
```diff
@@ -1747,13 +1820,16 @@ void esbmc_parseoptionst::help()
        "\nIncremental BMC\n"
        " --falsification              incremental loop unwinding for bug "
        "searching\n"
-       " --incremental-bmc            incremental loop unwinding verification\n"
+       " --incremental-bmc            incremental loop unwinding "
+       "verification\n"
```
Comment (mikhailramalho) [inline src/esbmc/esbmc_parseoptions.cpp]: No need to change this line.

File: src/esbmc/esbmc_parseoptions.cpp [hunk 11]
```diff
@@ -1747,13 +1820,16 @@ void esbmc_parseoptionst::help()
...
        " --termination                incremental loop unwinding assertion "
        "verification\n"
        " --k-step nr                  set k increment (default is 1)\n"
-       " --max-k-step nr              set max number of iteration (default is "
+       " --max-k-step nr              set max number of iteration (default "
+       "is "
```
Comment (mikhailramalho) [inline src/esbmc/esbmc_parseoptions.cpp]: No need to change this line.

File: src/esbmc/esbmc_parseoptions.cpp [hunk 12]
```diff
@@ -1747,13 +1820,16 @@ void esbmc_parseoptionst::help()
...
        "50)\n"
-       " --unlimited-k-steps          set max number of iteration to UINT_MAX\n"
+       " --unlimited-k-steps          set max number of iteration to "
+       "UINT_MAX\n"
```
Comment (mikhailramalho) [inline src/esbmc/esbmc_parseoptions.cpp]: No need to change this line.

File: src/esbmc/esbmc_parseoptions.cpp [hunk 13]
```diff
@@ -1763,7 +1839,8 @@ void esbmc_parseoptionst::help()
        " --cvc                        use CVC4\n"
        " --yices                      use Yices\n"
        " --bv                         use solver with bit-vector arithmetic\n"
-       " --ir                         use solver with integer/real arithmetic\n"
+       " --ir                         use solver with integer/real "
+       "arithmetic\n"
```
Comment (mikhailramalho) [inline src/esbmc/esbmc_parseoptions.cpp]: No need to change this line.

File: src/esbmc/esbmc_parseoptions.cpp [hunk 14]
```diff
@@ -1779,7 +1856,8 @@ void esbmc_parseoptionst::help()
        "                              SMT floating-point theory)\n"
        "--tuple-node-flattener        encode tuples using our tuple to node "
        "API\n"
-       "--tuple-sym-flattener         encode tuples using our tuple to symbol "
+       "--tuple-sym-flattener         encode tuples using our tuple to "
+       "symbol "
```
Comment (mikhailramalho) [inline src/esbmc/esbmc_parseoptions.cpp]: No need to change this line.

File: src/esbmc/esbmc_parseoptions.cpp [hunk 15]
```diff
@@ -1819,42 +1897,61 @@ void esbmc_parseoptionst::help()
        " --forward-condition          check the forward condition\n"
        " --inductive-step             check the inductive step\n"
        " --k-induction                prove by k-induction \n"
-       " --k-induction-parallel       prove by k-induction, running each step "
+       " --k-induction-parallel       prove by k-induction, running each "
+       "step "
```
Comment (mikhailramalho) [inline src/esbmc/esbmc_parseoptions.cpp]: No need to change this line.

File: src/esbmc/esbmc_parseoptions.cpp [hunk 16]
```diff
@@ -1819,42 +1897,61 @@ void esbmc_parseoptionst::help()
...
        "on a separate\n"
        "                              process\n"
        " --k-step nr                  set k increment (default is 1)\n"
-       " --max-k-step nr              set max number of iteration (default is "
+       " --max-k-step nr              set max number of iteration (default "
+       "is "
```
Comment (mikhailramalho) [inline src/esbmc/esbmc_parseoptions.cpp]: No need to change this line.

File: src/esbmc/esbmc_parseoptions.cpp [hunk 17]
```diff
@@ -1819,42 +1897,61 @@ void esbmc_parseoptionst::help()
...
        "50)\n"
-       " --unlimited-k-steps          set max number of iteration to UINT_MAX\n"
+       " --unlimited-k-steps          set max number of iteration to "
+       "UINT_MAX\n"
```
Comment (mikhailramalho) [inline src/esbmc/esbmc_parseoptions.cpp]: No need to change this line.

File: src/esbmc/esbmc_parseoptions.cpp [hunk 18]
```diff
@@ -1819,42 +1897,61 @@ void esbmc_parseoptionst::help()
...
        " --show-cex                   print the counter-example produced by "
        "the inductive step\n"
 
        "\nScheduling approaches\n"
        " --schedule                   use schedule recording approach \n"
-       " --round-robin                use the round robin scheduling approach\n"
+       " --round-robin                use the round robin scheduling "
+       "approach\n"
```
Comment (mikhailramalho) [inline src/esbmc/esbmc_parseoptions.cpp]: No need to change this line.

File: src/esbmc/esbmc_parseoptions.cpp [hunk 19]
```diff
@@ -1819,42 +1897,61 @@ void esbmc_parseoptionst::help()
...
        " --time-slice nr              set the time slice of the round robin "
        "algorithm\n"
        "                              (default is 1) \n"
 
        "\nConcurrency checking\n"
+       " --incremental-cb             incremental context-bound "
+       "verification\n"
```
Comment (mikhailramalho) [inline src/esbmc/esbmc_parseoptions.cpp:1913]: Check if the line fits the 80 columns.

File: src/esbmc/esbmc_parseoptions.cpp [hunk 20]
```diff
@@ -1819,42 +1897,61 @@ void esbmc_parseoptionst::help()
...
        " --context-bound nr           limit number of context switches for "
        "each thread \n"
        " --state-hashing              enable state-hashing, prunes duplicate "
        "states\n"
        " --no-por                     do not do partial order reduction\n"
        " --all-runs                   check all interleavings, even if a bug "
        "was already found\n"
+       " --initial-context-bound nr   set the initial context-bound for "
+       "incremental verification (default "
+       "is 2)\n"
```
Comment (mikhailramalho) [inline src/esbmc/esbmc_parseoptions.cpp]: This looks broken, check if the line fits the 80 columns.

File: src/esbmc/esbmc_parseoptions.cpp [hunk 21]
```diff
@@ -1819,42 +1897,61 @@ void esbmc_parseoptionst::help()
...
+       " --context-bound-step nr      set k context bound increment (default "
+       "is 5)\n"
+       " --max-context-bound nr       set max number of context-bound "
+       "(default "
+       "is "
+       "15)\n"
```
Comment (mikhailramalho) [inline src/esbmc/esbmc_parseoptions.cpp]: This looks broken, check if the line fits the 80 columns.

File: src/esbmc/esbmc_parseoptions.cpp [hunk 22]
```diff
@@ -1819,42 +1897,61 @@ void esbmc_parseoptionst::help()
...
+       " --unlimited-context-bound    set max number of context bounds to "
+       "UINT_MAX\n"
```
Comment (mikhailramalho) [inline src/esbmc/esbmc_parseoptions.cpp:1928]: Check if the line fits the 80 columns.

File: src/esbmc/esbmc_parseoptions.cpp [hunk 23]
```diff
@@ -1819,42 +1897,61 @@ void esbmc_parseoptionst::help()
...
 
        "\nMiscellaneous options\n"
-       " --memlimit                   configure memory limit, of form \"100m\" "
+       " --memlimit                   configure memory limit, of form "
+       "\"100m\" "
```
Comment (mikhailramalho) [inline src/esbmc/esbmc_parseoptions.cpp]: No need to change this line.

File: src/esbmc/esbmc_parseoptions.cpp [hunk 24]
```diff
@@ -1819,42 +1897,61 @@ void esbmc_parseoptionst::help()
...
        "or \"2g\"\n"
        " --timeout                    configure time limit, integer followed "
        "by {s,m,h}\n"
        " --memstats                   print memory usage statistics\n"
        " --no-simplify                do not simplify any expression\n"
        " --no-propagation             disable constant propagation\n"
        " --enable-core-dump           do not disable core dump output\n"
-       " --interval-analysis          enable interval analysis and add assumes "
+       " --interval-analysis          enable interval analysis and add "
+       "assumes "
```
Comment (mikhailramalho) [inline src/esbmc/esbmc_parseoptions.cpp]: No need to change this line.

File: src/esbmc/esbmc_parseoptions.cpp [hunk 25]
```diff
@@ -1114,6 +1117,69 @@ int esbmc_parseoptionst::doit_incremental()
...
+
+  bool unknown = true;
+
+  for(unsigned int context_bound = initial_context_bound;
+      context_bound <= max_context_bound;
+      context_bound += context_bound_inc)
+  {
+    char str[5];
+    snprintf(str, sizeof(str), "%d", context_bound);
+    opts.set_option("context-bound", str);
```
Comment (mikhailramalho) [inline src/esbmc/esbmc_parseoptions.cpp]: why are you using snprintf here? Check how we set the unwind option in methods like `do_base_case`.

**Review by mikhailramalho — CHANGES_REQUESTED** [review]
Looks almost fine. I didn't understand why you changed several lines in the option descriptions: if it was you IDE, check if the max column is set to less than 80.

File: src/esbmc/esbmc_parseoptions.cpp [hunk 26]
```diff
@@ -1837,13 +1911,24 @@ void esbmc_parseoptionst::help()
...
        " --context-bound nr           limit number of context switches for "
        "each thread \n"
        " --state-hashing              enable state-hashing, prunes duplicate "
        "states\n"
        " --no-por                     do not do partial order reduction\n"
        " --all-runs                   check all interleavings, even if a bug "
        "was already found\n"
+       " --initial-context-bound nr   set the initial context-bound for "
+       "incremental verification (default "
+       "is 2)\n"
```
Comment (mikhailramalho) [inline src/esbmc/esbmc_parseoptions.cpp]: the `is 2)` still looks broken.

File: src/esbmc/esbmc_parseoptions.cpp [hunk 27]
```diff
@@ -1837,13 +1911,24 @@ void esbmc_parseoptionst::help()
...
+       " --context-bound-step nr      set k context bound increment (default "
+       "is 5)\n"
+       " --max-context-bound nr       set max number of context-bound (default "
+       "is "
+       "15)\n"
```
Comment (mikhailramalho) [inline src/esbmc/esbmc_parseoptions.cpp]: the `15)` still looks broken.

File: src/esbmc/esbmc_parseoptions.cpp [hunk 28]
```diff
@@ -1114,6 +1117,77 @@ int esbmc_parseoptionst::doit_incremental()
...
+  // If we have explored all context bounds
+  // and did not find a property violation or
+  // the unwinding assertion does not hold,
+  // then we report verification unknown.
+  if(unknown)
+  {
+    status("Unable to prove or falsify the program, giving up.");
+    status("VERIFICATION UNKNOWN");
+  }
+
```
Comment (mikhailramalho) [inline src/esbmc/esbmc_parseoptions.cpp]: The use of the `unknown` variable still looks confusing. 

We are repeatedly calling the forward condition step, without changing the number of loop unwinding. Could you add an example where changing the number of context switches makes the forward condition report the wrong result?

Furthermore, we only set `unknown` to false if the forward condition is false, if it is true, shouldn't we set `unknown` to true? I mean, if for context bound = 2 the forward condition set `unknown` to false, but in context bound = 3 the forward condition doesn't hold, we are still reporting that the verification is not unknown.

Either unknown should be part of the branch condition to call the forward condition or the forward condition should be performed after we finished incrementing the context bound.

**Review by mikhailramalho — CHANGES_REQUESTED** [review]
Overall, I didn't like this approach because it is not compatible with incremental bmc, e.g., if the user sets --incremental-bmc, the incremental context-bound will be ignored.

I think the loop to increment the context bound should be moved the methods that call each step `do_inductive_step`, `do_forward_condition` and `do_base_case`).

**Review by mikhailramalho — COMMENTED** [review]
Much better! We are almost there:
1. Could you add the same logic to the steps in the k-induction parallel?
2. Now that the incremental-context-bound is not incompatible with the incremental-bmc, we should support cases where the user did not set the unwind or set it to a fixed value: this can be done by adding the proper logic at the end of the `doit()` method, surrounding: 
```
  // do actual BMC
  bmct bmc(goto_functions, opts, context, ui_message_handler);
  set_verbosity_msg(bmc);
  return do_bmc(bmc);
}
```

**Comment by lucasccordeiro** [comment]
@fatimahkj: Could you please take a look at this PR?

**Comment by lucasccordeiro** [comment]
@kaled-alshmrany: This is the issue.

**Review by mikhailramalho — APPROVED** [review]
LGTM

**Review by feliperodri — APPROVED** [review]
LGTM too.

**Pull request merged**